solutions/gold/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,347 @@
+
+**In summary, the transitions to $a_i$ are all states $\texttt{dp3}[x][\texttt{r}(a_i, b)][k_i - \texttt{bc}[x][\texttt{l}(a_i, b)]]$ where $x \in [0, 2^b)$**.
+
+Because the only variable in the state is $x$, the complexity of transition is the number of options for $x$ which is $\boxed{2^b}$.
+
+#### Affected States:
+We now have the length of the longest LBS that ends at $a_i$. What $\texttt{dp3}[x][y][k]$ states does this affect?
+
+Starting with the obvious, $x = \texttt{l}(a_i, b)$. We also have the restriction that $\texttt{r}(a_i, b) \in \texttt{trans}_k(y) \Rightarrow k = \texttt{bc}[\texttt{r}(a_i, b)][y]$. Because $\texttt{r}(a_i, b)$ is a constant, $k$ depends solely on $y$.
+
+**In summary, the states affected by $a_i$ are all states $\texttt{dp}3[\texttt{l}(a_i, b)][y][\texttt{bc}[\texttt{r}(a_i, b)][y]]$**.
+
+$y$ has to have the same amount of bits as $\texttt{r}(a_i, b)$ which has $\log_2{M} - b$ bits. Therefore, $y$ has $2^{\log_2M - b} = \boxed{\frac{M}{2^b}}$.
+
+To sum it all up, the complexity of transition is $\mathcal{O}(2^b)$ and the complexity of affected states is $\mathcal{O}(\frac{M}{2^b})$. To minimize the sum of these complexities, we should equate them. Because $M = 2 ^{20}$, we find the optimal value of $b$ to be $\boxed{10}$.
+
+### Implementation
+In the implementation below:
+- $\texttt{l}(x)$ denotes the 10 leftmost bits of $x$
+- $\texttt{r}(x)$ denotes the 10 righmost bits of $x$
+- $\texttt{dp}[i][j][k]$: what was previously referred to as $\texttt{dp3}$
+
+**Time Complexity:** $\mathcal{O}(N\sqrt{M})$
+
+<Spoiler title="Code">
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using pii = pair<int, int>;
+const int N = 1e5;
+const int B = 10;
+int bc[1 << B][1 << B];  // bc[i][j] = bit_count(i & j)
+struct state {
+	int length = 0, end = 0;
+	state() {}
+	state(int l, int e) : length(l), end(e) {}
+};
+state dp[1 << B][1 << B][B + 1];
+
+int main() {
+	// preprocess:
+	for (int i = 0; i < (1 << B); i++) {
+		for (int j = 0; j < (1 << B); j++) {
+			bc[i][j] = __builtin_popcount(i & j);
+		}
+	}
+
+	int n;
+	cin >> n;
+	vector<int> a(n);
+	for (int i = 0; i < n; i++) cin >> a[i];
+	vector<int> k(n);
+	for (int i = 0; i < n; i++) cin >> k[i];
+	int ans = 1, best_i = 0;  // best_i: best index to end subsequence at
+	vector<int> prv(n);       // prv[i]: best index to include before i
+	iota(prv.begin(), prv.end(), 0);  // prv[i] = i
+
+	for (int i = 0; i < n; i++) {
+		int l = a[i] >> 10, r = a[i] % (1 << 10);  // l(a[i]), r(a[i])
+		int lbs = 1;  // length of longest lbs that ends at i
+
+		// enumerate all possibilities for l(prev_num)
+		for (int j = 0; j < (1 << B); j++) {
+			// here, we use the fact that
+			// bc[x][y] = bc[l(x)][l(y)] + bc[r(x)][r(y)],
+			// or bc[r(x)][r(y)] = bc[x][y] - bc[l(x)][l(y)]
+
+			// required value of bc[r(x)][r(y)]
+			int needed = k[i] - bc[j][l];
+			if (needed < 0 || needed > 10) continue;
```

Comment:
braces
also where'd the 10 come from

Reply:
if it's defined in terms of a problem-wise constant, set it w/ `const` and then do it